[cxx-interop] Allow importing templated functions when template args do not appear in function signature

guitard0g opened the pull request:
Currently if we try to import a templated function when the template parameters do not appear in the function signature, the corresponding Swift function will have extra unused generic parameters, which means we can't call those functions. This patch fixes the problem by making those template params into explicit metatype function parameters to the corresponding Swift function. E.g.
```
template <typename T>
void foo()
```
becomes:
```
func foo<T>(T: T.Type) 
```
instead of:
```
func foo<T>() // can't be called
```
This extra argument is only used to deduce the generic type needed to specialize the C++ function. The corresponding function call made under the hood will cut out the extra argument and forward along the proper parameters to the C++ function.

Commit:
Allow importing templated functions when template args do not appear
in the function signature by adding explicit metatype parameters to
the function signature.

guitard0g commented:
WIP because there seems to be some wonkiness with how this interacts with default template arguments, still investigating a fix there.

zoecarver commented on lib/ClangImporter/ImporterImpl.h:
```diff
@@ -1560,6 +1560,9 @@ bool isSpecialUIKitStructZeroProperty(const clang::NamedDecl *decl);
 /// even if imported as a non-static member function.
 bool isImportedAsStatic(clang::OverloadedOperatorKind op);
 
+bool hasSameUnderlyingType(const clang::Type *a,
```
A small comment about what this does would be nice as it's not immediately obvious from the name.

zoecarver commented on lib/Sema/CSApply.cpp:
```diff
@@ -109,7 +109,7 @@ Solution::computeSubstitutions(GenericSignature sig,
 }
 
-static ConcreteDeclRef generateDeclRefForSpecializedCXXFunctionTemplate(
+static ValueDecl *generateDeclForSpecializedCXXFunctionTemplate(
```
I know I named this function, but it's kind of ugly, so while you're updating the signature maybe change it to `generateSpecializedCXXFunctionTemplate`? WDYT?

zoecarver commented on lib/Sema/CSApply.cpp:
```diff
@@ -176,7 +186,49 @@ static ConcreteDeclRef generateDeclRefForSpecializedCXXFunctionTemplate(
+// metatype arguments to aid in typechecking, but they shouldn't be forwarded to
+// the corresponding C++ function.
+static std::pair<BraceStmt *, bool>
+synthesizeForwardingThunkBody(AbstractFunctionDecl *afd, void *context) {
```
Not in this patch, but at some point we should move all synthesizers to their own file. They just take up space right now and clutter the files they're scattered across.

zoecarver commented on test/Interop/Cxx/templates/Inputs/template-type-parameter-not-in-signature.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef TEST_INTEROP_CXX_TEMPLATES_INPUTS_TEMPLATE_TYPE_PARAMETER_NOT_IN_SIGNATURE_H
+#define TEST_INTEROP_CXX_TEMPLATES_INPUTS_TEMPLATE_TYPE_PARAMETER_NOT_IN_SIGNATURE_H
+
```
These tests are great! Thanks for being thourough :)

Two suggestions: can we get a test where there are two template params and only one is used in the signature. And a test where there's a non-type-template-param (it doesn't matter how we import it, just make sure we don't crash). (Side note: at some point we'll probably want to do something similar to this patch but with non-type-template-params.)

zoecarver commented on test/Interop/Cxx/templates/template-type-parameter-not-in-signature.swift:
```diff
@@ -0,0 +1,22 @@
+  templateTypeParamNotUsedInSignature(T: Int.self)
+  multiTemplateTypeParamNotUsedInSignature(T: Float.self, U: Int.self)
+  multiTemplateTypeParamNotUsedInSignatureWithUnrelatedParams(1, 1, T: Int32.self, U: Int.self)
+  let _: Int = templateTypeParamUsedInReturnType(10)
```
Can you add an `expectEquals` somewhere in both of these tests. Otherwise (very theoretically), they might just get optimized away. Also it's not a bad idea to just test that the correct thing is returned and there wasn't an argument mismatch or something.

zoecarver reviewed (approved):
What a great patch! Thank you, Josh.

I really like how you commented all the code you added so it's super easy to understand and follow. Also, thanks for factoring out that lambda into a general helper function. 

I have one or two small nits, but overall, this patch seems good to me :)

zoecarver commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -1980,6 +1980,49 @@ ParameterList *ClangImporter::Implementation::importFunctionParameterList(
+                           bool shouldCheckResultType) -> bool {
+    auto paramDecl = genericParam->getClangDecl();
+    auto templateTypeParam = cast<clang::TemplateTypeParmDecl>(paramDecl);
+    // TODO: This won't work when we support importing dependent types. we'll
```
Oops, meant to actually give you the SR. Just update the TODO to read `TODO(SR-13809):`. This way we can grep for the SR and find all the places we need to update whenever we add support for dependent types.

zoecarver commented:
@guitard0g let me know if you need help running the bots.

egorzhdan commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -1980,6 +1980,49 @@ ParameterList *ClangImporter::Implementation::importFunctionParameterList(
+  
+  // Make sure all generic parameters are accounted for in the function signature.
+  for (auto genericParam : genericParams) {
+    if (genericParamTypeUsedInSignature(genericParam, resultType->hasTypeParameter()))
```
This currently crashes when importing `std` because `resultType` might be null. You can fix that by replacing `resultType->hasTypeParameter()` with `resultType && resultType->hasTypeParameter()`.

zoecarver replied:
Good catch. And this will also make it easier to add support for lazily loading result types.

egorzhdan reviewed (commented):
This approach looks great to me! I only noticed one small issue that prevents the C++ stdlib from being imported.

guitard0g commented:
@swift-ci please test

guitard0g commented:
@swift-ci please test

guitard0g commented:
@egorzhdan @zoecarver Thanks for the review! I believe I've addressed all the comments in the most recent revision. It changed a decent bit because I realized there were some quirks with making sure we manually invoke the generic substitutions on the signature of the synthesized thunk. Should be ready for another round of reviews.

zoecarver commented on lib/Sema/CSApply.cpp:
```diff
@@ -127,22 +129,49 @@ static ConcreteDeclRef generateDeclRefForSpecializedCXXFunctionTemplate(
+static ValueDecl *generateSpecializedCXXFunctionTemplate(
+    ASTContext &ctx, AbstractFunctionDecl *oldDecl, SubstitutionMap subst,
+    clang::FunctionDecl *specialized) {
+  auto newFnTypeAndParams = substituteFunctionTypeAndParamList(ctx, oldDecl, subst);
```
I wish we could use `auto [a, b] = ` 😢

zoecarver commented on lib/Sema/CSApply.cpp:
```diff
@@ -127,22 +129,49 @@ static ConcreteDeclRef generateDeclRefForSpecializedCXXFunctionTemplate(
+      // exist in the resulting specialized FuncDecl. Note that this doesn't
+      // affect constructors because all template params for a constructor
+      // must be in the function signature by design.
+      continue;
```
I think various methods/static methods might take a meta type. We need to verify if this will break that.

zoecarver commented on lib/Sema/CSApply.cpp:
```diff
@@ -109,16 +109,18 @@ Solution::computeSubstitutions(GenericSignature sig,
-    ASTContext &ctx, AbstractFunctionDecl *oldDecl, SubstitutionMap subst,
-    clang::FunctionDecl *specialized) {
+// Derive a concrete function type for fdecl by substituting the generic args
+// and use that to derive the corresponding return type and parameter list.
```
```suggestion
// and use that to derive the corresponding function type and parameter list.
```

zoecarver commented on lib/Sema/CSApply.cpp:
```diff
@@ -176,7 +205,49 @@ static ConcreteDeclRef generateDeclRefForSpecializedCXXFunctionTemplate(
-  return ConcreteDeclRef(newFnDecl);
+  return newFnDecl;
+}
+
```
Remove space

zoecarver commented on lib/Sema/CSApply.cpp:
```diff
@@ -215,13 +286,35 @@ Solution::resolveConcreteDeclRef(ValueDecl *decl,
         decl->getASTContext(), cast<AbstractFunctionDecl>(decl), subst, newFn);
+    if (auto fn = dyn_cast<FuncDecl>(decl)) {
+      if (newFn->getNumParams() != fn->getParameters()->size()) {
+        // We probably added additional metatype parameters to aid template
```
```suggestion
        // We added additional metatype parameters to aid template
```

zoecarver reviewed (approved):
🚢 it

(I left three very small nits. All are optional.)

zoecarver commented:
> Swift Test Windows Platform — Build failed. 

Now you see why I'm so excited about #36082 :P

egorzhdan reviewed (approved):
Looks great!
